### ISSUE TITLE
sometime throws HttpParsingError

### ISSUE THREAD

**Issue by Sunda001** [sonic182/aiosonic#262]
so i do a request to an API but it sometimes raise these errors:

```
2021-06-27T12:02:51.478515+00:00 app[worker.1]:     return await self.request(url,
2021-06-27T12:02:51.478515+00:00 app[worker.1]:   File "/app/.heroku/python/lib/python3.9/site-packages/aiosonic/__init__.py", line 738, in request
2021-06-27T12:02:51.478516+00:00 app[worker.1]:     response = await wait_for(
2021-06-27T12:02:51.478516+00:00 app[worker.1]:   File "/app/.heroku/python/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
2021-06-27T12:02:51.478516+00:00 app[worker.1]:     return fut.result()
2021-06-27T12:02:51.478516+00:00 app[worker.1]:   File "/app/.heroku/python/lib/python3.9/site-packages/aiosonic/__init__.py", line 433, in _do_request
2021-06-27T12:02:51.478518+00:00 app[worker.1]:     async with (await connector.acquire(*args)) as connection:
2021-06-27T12:02:51.478518+00:00 app[worker.1]:   File "/app/.heroku/python/lib/python3.9/site-packages/aiosonic/connectors.py", line 59, in acquire
2021-06-27T12:02:51.478518+00:00 app[worker.1]:     raise HttpParsingError('missing hostname')
2021-06-27T12:02:51.478518+00:00 app[worker.1]: aiosonic.exceptions.HttpParsingError: missing hostname
```

aiosonic version: **0.10.1**

**Comment by sonic182** [comment]
missing hostname exception mmm... when parsing url it is not getting hostname, it parses url with this function https://github.com/sonic182/aiosonic/blob/a04f611da2a584fcbaac9fd8b7ebab2e1d1c7437/aiosonic/__init__.py#L71 can you check the url you're using to call the request?

**Comment by Sunda001** [comment]
hmm sometimes it works fine but sometimes it throws that errors, this kind of errors is always coming in aiosonic 0.9.7 but its not coming frequently after i upgraded it to the latest version, but still that kind of errors is coming sometimes

**Comment by Sunda001** [comment]
the params of url is containing  '[]' like this:
`params={'foo': '[this]'}`
maybe it could be the issues?

**Comment by sonic182** [comment]
I don't think so, gonna add some debug logs for next version (whenever I got time) so it is easier to debug this kind of errors.

Try to log the arguments you're using to call the request please :)

**Comment by Sunda001** [comment]
what do you mean by 'log the arguments' ?

**Comment by sonic182** [comment]
```python

def something():
    url = "https://postman-echo.com/post"
    posted_data = {'foo': 'bar'}
    # log stuffs
    logger.info(f'url={url} posted_data={str(posted_data)}')
    response = await client.post(url, data=posted_data)
```

**Comment by Sunda001** [comment]
ah okay

**Comment by Sunda001** [comment]
now i get it again...
here is the full traceback and log of the url
```
url=https://vm.tiktok.com/ZSJVWrMAB/
Traceback (most recent call last):
  File "/app/tiktokDL.py", line 98, in _redr
    res = await tt_tools.request.get(foo, headers={'User-Agent': self.user_agent})
  File "/app/.heroku/python/lib/python3.9/site-packages/aiosonic/__init__.py", line 563, in get
    return await self.request(url,
  File "/app/.heroku/python/lib/python3.9/site-packages/aiosonic/__init__.py", line 743, in request
    response = await wait_for(
  File "/app/.heroku/python/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
  File "/app/.heroku/python/lib/python3.9/site-packages/aiosonic/__init__.py", line 459, in _do_request
    response._set_response_initial(
  File "/app/.heroku/python/lib/python3.9/site-packages/aiosonic/__init__.py", line 158, in _set_response_initial
    raise HttpParsingError('response line parsing error')
aiosonic.exceptions.HttpParsingError: response line parsing error
```

but for the second try it works, but sometimes throws it again

**Comment by sonic182** [comment]
Mmm something weird, gonna improve that exception next release

This sample code works, with the url you have in the backtrace, you can try add sentry to your project, it may show the variables values in those steps of the error.

```python
import asyncio
from aiosonic import HTTPClient


async def main():

    async with HTTPClient() as client:
        res = await client.get('https://vm.tiktok.com/ZSJVWrMAB/', follow=True)
        print(res.status_code, await res.text())



asyncio.run(main())
```

**Comment by Sunda001** [comment]
what if i using on class?
can i do this instead of using context manager on every function
```
class Foo:
    def __init__(self):
        self._request = HTTPClient(verify_ssl=False)

    async def home(self):
        await self._request.get(...)

    async def timeline(self):
        await self._request.get(...)

    def main(self):
        await self.home()
        await self.timeline()

    def __del__(self):
        await self._request.close()
```

**Comment by sonic182** [comment]
That way you don't take advantage of keep alive feature, better if you use one instance of `HTTPClient` for all requests, you could wrap it in a singleton class or just a global variable, however fits better to your code

**Comment by sonic182** [comment]
@Sunda001 you can try 0.11.1 and activate debug log https://aiosonic.readthedocs.io/en/latest/examples.html#debug-log if you have not too much traffic in your app

I did improve the `aiosonic.exceptions.HttpParsingError` to it display the data read by the socket that is bad then raising that exception.

**Comment by Sunda001** [comment]
full traceback:
```
url=https://vm.tiktok.com/ZSJqYw3rv/
Traceback (most recent call last):
  File "/app/tiktokDL.py", line 98, in _redr
    res = await tt_tools.request.get(url, headers={'User-Agent': self.user_agent})
  File "/app/.heroku/python/lib/python3.9/site-packages/aiosonic/__init__.py", line 572, in get
    return await self.request(url,
  File "/app/.heroku/python/lib/python3.9/site-packages/aiosonic/__init__.py", line 752, in request
    response = await wait_for(
  File "/app/.heroku/python/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
  File "/app/.heroku/python/lib/python3.9/site-packages/aiosonic/__init__.py", line 468, in _do_request
    response._set_response_initial(
  File "/app/.heroku/python/lib/python3.9/site-packages/aiosonic/__init__.py", line 143, in _set_response_initial
    raise HttpParsingError(f'response line parsing error: {data}')
aiosonic.exceptions.HttpParsingError: response line parsing error: b''
```

logging:
````
2021-07-01T06:33:36.522243+00:00 app[worker.1]: X-Tt-Logid=2021070106333601018907401642545260
2021-07-01T06:33:36.522244+00:00 app[worker.1]: x-tt-trace-host=01b22ef7dd313e5116960c9bda12d666fdb940c13f7e87334ebe32540fd6913c6417a81f84e4a1b6eab5429d347dfc5914ce904b57b5a2a101c44b05418efb5424617614d80ee0578896d43c262b4557d54e0fcae9b58bdb9449d24b74695583021d27061d21b0310ff17eba9395f0fdba
2021-07-01T06:33:36.522244+00:00 app[worker.1]: X-Origin-Response-Time=5,23.217.129.77
2021-07-01T06:33:36.522245+00:00 app[worker.1]: X-Akamai-Request-ID=5a7f714.11c2fa1
2021-07-01T06:33:36.522245+00:00 app[worker.1]: Date=Thu, 01 Jul 2021 06:33:36 GMT
2021-07-01T06:33:36.522245+00:00 app[worker.1]: X-Cache=TCP_MISS from a2-21-33-86.deploy.akamaitechnologies.com (AkamaiGHost/10.4.0.3-34328257) (-)
2021-07-01T06:33:36.522246+00:00 app[worker.1]: Connection=keep-alive
2021-07-01T06:33:36.522246+00:00 app[worker.1]: X-Cache-Remote=TCP_MISS from a23-217-129-77.deploy.akamaitechnologies.com (AkamaiGHost/10.4.0.3-34328257) (-)
2021-07-01T06:33:36.522246+00:00 app[worker.1]: X-Parent-Response-Time=85,2.21.33.86
2021-07-01T06:33:36.522248+00:00 app[worker.1]: x-tt-trace-tag=id=16;cdn-cache=miss;type=dyn
2021-07-01T06:33:36.522249+00:00 app[worker.1]: Server-Timing=cdn-cache; desc=MISS, edge; dur=81, origin; dur=5
2021-07-01T06:33:36.522249+00:00 app[worker.1]: Server-Timing=inner; dur=0---
2021-07-01T06:45:04.442842+00:00 app[worker.1]: GET /ZSJqYw3rv/ HTTP/1.1
2021-07-01T06:45:04.442849+00:00 app[worker.1]: HOST: vm.tiktok.com
2021-07-01T06:45:04.442850+00:00 app[worker.1]: Connection: keep-alive
2021-07-01T06:45:04.442850+00:00 app[worker.1]: User-Agent: aioload/0.11.1
2021-07-01T06:45:04.442851+00:00 app[worker.1]: User-Agent: TikTok 16.0.16 rv:103005 (iPhone; iOS 11.1.4; en_EN) Cronet
2021-07-01T06:45:04.442852+00:00 app[worker.1]: ---
```

**Comment by sonic182** [comment]
seems that it is not reading from the socket any data when it should get the first status line. The only weird stuff is the user agent header repeated, seems that It is concatenating it instead of replacing the User-Agent header. Gonna fix that whenever I get time

**Comment by sonic182** [comment]
Please @Sunda001 can you try with 0.11.2?

Thanks!

**Comment by sonic182** [comment]
The real fix should be in 0.11.3, not in 0.11.12 CC @Sunda001

**Comment by sonic182** [comment]
Please @Sunda001 if you get any error with `aiosonic>=0.11.3`, please don't hesitate to comment here again, I'm gonna close the issue for now
